File: 16-Proxy/src/go/main.go
package main

import "fmt"

func main() {
	fmt.Println("=== Proxy Pattern Demo (Virtual / Lazy Loading, Go) ===")
	fmt.Println()

	// Creating the proxies is cheap — no disk load happens here
	fmt.Println("Creating image proxies (no loading expected):")
	var photo1 Image = NewProxyImage("vacation.jpg")
	var photo2 Image = NewProxyImage("portrait.jpg")

	// The heavy load happens only when Display() is first called
	fmt.Println("\nFirst display of photo1 (triggers load):")
	photo1.Display()

	fmt.Println("\nSecond display of photo1 (already loaded, no reload):")
	photo1.Display()

	fmt.Println("\nphoto2 was never displayed — so it was never loaded.")
	fmt.Println("That is the point of a virtual proxy: pay the cost only if used.")
	_ = photo2
}

File: 16-Proxy/src/rust/main.rs
mod image;
mod real_image;
mod proxy_image;

use image::Image;
use proxy_image::ProxyImage;

fn main() {
    println!("=== Proxy Pattern Demo (Virtual / Lazy Loading, Rust) ===\n");

    // Creating the proxies is cheap — no disk load happens here
    println!("Creating image proxies (no loading expected):");
    let photo1: Box<dyn Image> = Box::new(ProxyImage::new("vacation.jpg"));
    let _photo2: Box<dyn Image> = Box::new(ProxyImage::new("portrait.jpg"));

    // The heavy load happens only when display() is first called
    println!("\nFirst display of photo1 (triggers load):");
    photo1.display();

    println!("\nSecond display of photo1 (already loaded, no reload):");
    photo1.display();

    println!("\nphoto2 was never displayed — so it was never loaded.");
    println!("That is the point of a virtual proxy: pay the cost only if used.");
}
